common/meta/types/src/user_auth.rs

```diff
@@ -11,17 +11,237 @@
+            },
+        }
+    }
+
+    pub fn get_password(&self) -> Option<Vec<u8>> {
+        match self {
+            AuthInfo::Password {
+                password: p,
+                password_type: _,
+            } => Some(p.to_vec()),
+            _ => None,
+        }
+    }
+
+    pub fn get_password_type(&self) -> Option<PasswordType> {
+        match self {
+            AuthInfo::Password {
+                password: _,
+                password_type: t,
+            } => Some(*t),
+            _ => None,
+        }
+    }
+
+    fn decode_double_sha1(
+        salt: &[u8],
+        input: &[u8],
+        user_password: &[u8],
+    ) -> Result<Vec<u8>, ErrorCode> {
+        let double_sha1 = calc_sha1(&calc_sha1(user_password)[..]);
+        // SHA1( password ) XOR SHA1( "20-bytes random data from server" <concat> SHA1( SHA1( password ) ) )
+        let mut m = sha1::Sha1::new();
+        m.update(salt);
+        m.update(&double_sha1[..]);
+
+        let result = m.digest().bytes();
+        if input.len() != result.len() {
+            return Err(ErrorCode::SHA1CheckFailed("SHA1 check failed"));
+        }
+        let mut s = Vec::with_capacity(result.len());
+        for i in 0..result.len() {
+            s.push(input[i] ^ result[i]);
+        }
+        Ok(s)
+    }
+
+    pub fn auth_mysql(&self, password_input: &[u8], salt: &[u8]) -> Result<bool, ErrorCode> {
+        match self {
+            AuthInfo::None => Ok(true),
+            AuthInfo::Password {
+                password: p,
+                password_type: t,
+            } => {
+                match t {
+                    PasswordType::PlainText => Ok(p == password_input),
+                    PasswordType::DoubleSha1 => {
+                        let password_sha1 = AuthInfo::decode_double_sha1(salt, password_input, p)?;
+                        // TODO(youngsofun): we should store double_sha1(password) later
+                        Ok(double_sha1(p) == calc_sha1(&password_sha1))
+                    }
+                    PasswordType::Sha256 => Err(ErrorCode::AuthenticateFailure(
+                        "login with sha256_password user for mysql protocol not supported yet.",
+                    )),
+                }
+            }
+        }
+    }
+
+    // for clickhouse and http basic
+    pub fn auth_password_plaintext(
+        &self,
+        password_input_plaintext: &[u8],
+    ) -> Result<bool, ErrorCode> {
+        match self {
+            AuthInfo::None => Ok(true),
+            AuthInfo::Password {
+                password: p,
+                password_type: _t,
+                // TODO(youngsofun): we should store password hash later
+            } => Ok(p == password_input_plaintext),
+        }
+    }
+}
+
+#[derive(Clone, serde::Serialize, serde::Deserialize, Debug, PartialEq)]
+pub struct AuthInfoArgs {
+    pub arg_with: Option<AuthType>,
+    pub arg_by: Option<String>,
+}
```

Comment:
can we merge AuthInfoArgs and AuthInfo into a unified struct? q.q

i think we can dispatch the auth methods by name in the sql parser.

Reply:
I tried it at first but did not find a good way.


 there are tiny differences between "create user" and "alter user":

create user:  
- users do not need to provide an arg as long as **it has a reasonable default value**

alter user:
- user users do not need to provide an arg as long as i**t makes sense to use the current value (just change the supplied parts)** 
- it has more situations
    - diff auth type may use diff args like as, by, server in click house 
   - a SQL may be valid by itself (from parser view), but invalid combined with the current auth type. 

 so for an AuthInfo::XXX, we may need a type AuthInfoArgs::XXX or r AuthInfo::XXX::Args, I think it is not necessary for now.


## this pr:
 
AuthInfo is stored in meta, as the final records. Most Fields in AuthInfo are not optional. 

create user:  
- parser:  SQL -> AuthInfoArgs -> AuthInfo
- analyzer use AuthInfo

alter user:
- parser:  SQL -> AuthInfoArgs(check grama of the SQL itself)
- analyzer make use of AuthInfoArgs according to current AuthInfo (may check futher)

Reply:
So, it seems we don't need a AuthInfoArg in meta here, it only has two fields, it would better to use function argument or others do that.
I am still confused on AuthInfo and AuthInfoArgs 🤔

Reply:
@BohuTANG. 

this is for extensions later.
 
currently,  it only has two fields now because we only implement password auth now. 
later,  diff auth type will use many diff arg names,  It may looks like 


```
pub struct AuthInfoArgs {
    pub arg_with: Option<AuthType>,
    pub arg_by: Option<String>,
    pub arg_as: Option<String>,   // for mysql style ladp
    pub arg_server: Option<String>, // for clickhouse  style ladp
    pub arg_realm: Option<String>,   // for cickhouase style kerberos
  ...
}
```


we can all use "BY":
some may need more than one args.
some may need other types like int.  
even for passwords, we may later add something like password_expired (which most auth types do not need)

with diff names and types, is easier for user to understand and for parser to check

----

passing them by function argument will make the argument list of all related functions too long and with many options fields. 

---

and in "alter user",  need to check it in 2 places:

```
create user yang IDENTIFIED WITH ldap SERVER 'server_name'
alter user yang IDENTIFIED SERVER 'server_name' //  ok
alter user yang IDENTIFIED by 'new_password' //  ok to the parser,  fail for interpreter 
```

----

with AuthInfoArgs,  the info in SQL info is collected to a struct,   ease the reading,checking and passing around.

 it can be seen as a direct result struct (AST) of a general parser (like antlr),  which is natural  and is not an extra work.

Reply:
Ok, thank you for explaining.

Reply:
> later, diff auth type will use many diff arg names, It may looks like
>
> ```
> pub struct AuthInfoArgs {
>    pub arg_with: Option<AuthType>,
>    pub arg_by: Option<String>,
>    pub arg_as: Option<String>,   // for mysql style ladp
>    pub arg_server: Option<String>, // for clickhouse  style ladp
>    pub arg_realm: Option<String>,   // for cickhouase style kerberos
>  ...
> }
> ```

i'm afraid this AuthInfoArgs may contains too many details on different protocols in one struct, the fields might overlap on different scienaries. maybe a enum is more suitable on repsenting different things?

or as @Xuanwo 's suggest https://github.com/multiformats/rust-multihash , is it possible to represent multiple auth data in one auth string? q.q

Reply:
> AuthInfoArgs is only an intermediate structure, not stored.

i think AuthInfoArgs is an struct mostly used in the ast layer, and it'd converted to AuthInfo in the interpreter or planner layer, and stores inside the metastore.

the `meta` mod mostly stores some seriaklizable objects in the metastore, how about move the AuthInfoArgs to the `ast` mod?

Reply:
> it possible to represent multiple auth data in one auth string? q.q

1. rust-multihash is good but only useful for password

> we can not all use "BY":
some may need more than one args.
some may need other types like int.
even for passwords, we may later add something like password_expired (which most auth types do not need)
with diff names and types, is easier for user to understand and for parser to check

Reply:
AuthInfoArgs  

> maybe a enum is more suitable on repsenting different things

I tried at first,  one problem is  "with" can be omitted,  so need to guess enum type from only "by xxx"
currently not sure "by" is only one auth type, depends on how we design the SQL.